# STAT 207 Lab Assignment 7 - Group Part


## <u>Case Study</u>: Impact of Genre on Song Metric Importance in Predicting Popularity

The dataset that we will analyze in this group assignment is comprised of information about the top 100 songs of each year on Spotify from 2010 to 2019. Each row contains the following information about a particular song.

* `title`	Song's Title
* `artist`	Song's artist
* `genre`	Genre of song
* `pop`	Popularity of the song (not a ranking)
* `bpm`	Beats Per Minute - The tempo of the song
* `nrgy`	Energy - How energetic the song is
* `dnce`	Danceability - How easy it is to dance to the song
* `dB`	Decibel - How loud the song is
* `live`	How likely the song is a live recording
* `val`	How positive the mood of the song is
* `dur`	Duration of the song
* `acous`	How acoustic the song is
* `spch`	The more the song is focused on spoken word


**This dataset has no missing values**.


## Research Goal

1. First you get to pick two popular song genres in this dataset.
2. Then, we'd like to determine and compare what the top three most important `song metrics` are when it comes to predicting the `popularity` of songs for these two different genres.

## Points

<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
    </tbody>
</table>



In [6]:
import pandas as pd

## 1. Dataset Exploration

### 1.1. Dataset

1. Read the `spotify_df.csv` file into a dataframe.
2. Show the first five rows.
3. Show how many rows are in the dataframe.

In [7]:
spotify_df = pd.read_csv('spotify_df.csv')
first_five_rows = spotify_df.head()
num_rows = len(spotify_df)
first_five_rows, num_rows

(                                           title      artist    top genre  \
 0                  STARSTRUKK (feat. Katy Perry)       3OH!3    dance pop   
 1                    My First Kiss (feat. Ke$ha)       3OH!3    dance pop   
 2                                I Need A Dollar  Aloe Blacc     pop soul   
 3  Airplanes (feat. Hayley Williams of Paramore)       B.o.B  atl hip hop   
 4              Nothin' on You (feat. Bruno Mars)       B.o.B  atl hip hop   
 
    popularity  bpm  nrgy  dnce  dB  live  val  dur  acous  spch  
 0          70  140    81    61  -6    23   23  203      0     6  
 1          68  138    89    68  -4    36   83  192      1     8  
 2          72   95    48    84  -7     9   96  243     20     3  
 3          80   93    87    66  -4     4   38  180     11    12  
 4          79  104    85    69  -6     9   74  268     39     5  ,
 1000)

### 1.2. Genres with Larger Sample Sizes

Suppose we would only like to consider genres in this dataset that have at least 20 songs. Create a **list** (or Index() object) of song genres that have at least 20 songs in this dataset.

In [8]:
genre_counts = spotify_df['top genre'].value_counts()

In [9]:
genres_with_min_20_songs = genre_counts[genre_counts >= 20].index.tolist()
genres_with_min_20_songs

['dance pop', 'pop', 'atl hip hop', 'art pop', 'hip hop', 'boy band']

### 1.3. Picking Two Genres

1. Next pick two genres from this list that you'd like to explore. 
2. Then create two dataframes from your original dataframe: a.) one that is comprised of songs songs from the first genre, b.) the other that is comprised of songs from the second genre.

In [10]:
dance_pop_df = spotify_df[spotify_df['top genre'] == 'dance pop']
pop_df = spotify_df[spotify_df['top genre'] == 'pop']

dance_pop_df_head = dance_pop_df.head()
pop_df_head = pop_df.head()

dance_pop_df_head, pop_df_head


(                           title           artist  top genre  popularity  bpm  \
 0  STARSTRUKK (feat. Katy Perry)            3OH!3  dance pop          70  140   
 1    My First Kiss (feat. Ke$ha)            3OH!3  dance pop          68  138   
 6           The Time (Dirty Bit)  Black Eyed Peas  dance pop          75  128   
 7                        Imma Be  Black Eyed Peas  dance pop          71   92   
 8            Talking to the Moon       Bruno Mars  dance pop          87  146   
 
    nrgy  dnce  dB  live  val  dur  acous  spch  
 0    81    61  -6    23   23  203      0     6  
 1    89    68  -4    36   83  192      1     8  
 6    81    82  -8    60   44  308      7     7  
 7    52    60  -7    31   41  258     18    37  
 8    59    50  -5    11    8  218     51     3  ,
                      title        artist top genre  popularity  bpm  nrgy  \
 54                  Misery      Maroon 5       pop          70  103    81   
 78        Back To December  Taylor Swift       p

## 2. First Genre Analysis

First, we'd like to fit a *type of* linear regression model that predicts a songs `popularity` using the following explanatory variables for `just your genre 1 songs`:
* `bpm`	Beats Per Minute - The tempo of the song
* `nrgy`	Energy - How energetic the song is
* `dnce`	Danceability - How easy it is to dance to the song
* `dB`	Decibel - How loud the song is
* `live`	How likely the song is a live recording
* `val`	How positive the mood of the song is
* `dur`	Duration of the song
* `acous`	How acoustic the song is
* `spch`	The more the song is focused on spoken word

HOWEVER, we'd ALSO like to determine HOW IMPORTANT each of these explanatory variables are when it comes to predicting song `popularity`.

### 2.1. Song Metric Standard Deviations

First, calculate the standard deviation of each of our song metric explanatory variables for `just your genre 1 songs`

* `bpm`	Beats Per Minute - The tempo of the song
* `nrgy`	Energy - How energetic the song is
* `dnce`	Danceability - How easy it is to dance to the song
* `dB`	Decibel - How loud the song is
* `live`	How likely the song is a live recording
* `val`	How positive the mood of the song is
* `dur`	Duration of the song
* `acous`	How acoustic the song is
* `spch`	The more the song is focused on spoken word

In [11]:
std_dance_pop = dance_pop_df[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']].std()
std_dance_pop

bpm      22.872566
nrgy     13.458405
dnce     11.469767
dB        1.567473
live     14.800015
val      21.373355
dur      40.193552
acous    15.207323
spch      7.303473
dtype: float64

### 2.2. Usability of an *Unscaled* Linear Regression Model

Suppose we were to just simply fit a linear regression model that predicts song `popularity` with your `first genre dataset` given the UNSCALED versions of our song metric explanatory variables. Would we be able to interpret the absolute value of the resulting slopes as an indication of how IMPORTANT the corresponding explanatory variables were when it comes to predicting `popularity` of songs from the `first genre dataset`? Explain.

No, we would not be able to interpret the absolute value of the resulting slopes as an indication of how important the corresponding explanatory variables were when it comes to predicting `popularity` of songs from the `first genre dataset` if we used the UNSCALED versions of our song metric explanatory variables. Here's why:

1. **Different Units and Ranges**: Each of the song metric explanatory variables has different units and ranges. For instance, `bpm` might range from 60 to 200, while `dB` might range only a few units. A small change in `dB` might have a significant impact on popularity, but the same absolute change in `bpm` might not.

2. **Magnitude of Coefficients**: In an unscaled linear regression model, the magnitude of the coefficients (slopes) is influenced by the scale of the explanatory variables. A variable with a larger scale (like `dur` which represents song duration in seconds) might have a smaller coefficient compared to a variable with a smaller scale. This doesn't mean the variable with the larger scale is less important; it's just a consequence of the different scales.

3. **Interpretation Difficulties**: Without scaling, a unit change in one variable might not be comparable to a unit change in another variable. For example, a one-unit change in `bpm` is not equivalent to a one-unit change in `dnce` in terms of their impact on the song's `popularity`.

### 2.3. z-Score Scaling the Explanatory Variables

Next, scale each of your explanatory variables in your `first genre dataset` using the z-score scaling method that we discussed in class.

In [13]:
columns_to_scale = ['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']
genre1_data = spotify_df[spotify_df['top genre'] == 'dance pop']

genre1_scaled = genre1_data.copy()
genre1_scaled[columns_to_scale] = (genre1_data[columns_to_scale] - genre1_data[columns_to_scale].mean()) / genre1_data[columns_to_scale].std()

genre1_scaled[columns_to_scale].head() 


bpm      nrgy      dnce        dB      live       val       dur  \
0  0.871624  0.568284 -0.527703 -0.546074  0.243318 -1.476195 -0.374641   
1  0.784183  1.162708  0.082597  0.729865  1.121695  1.331038 -0.648317   
6  0.346978  0.568284  1.303197 -1.822013  2.743315 -0.493664  2.237718   
7 -1.226960 -1.586503 -0.614889 -1.184043  0.783858 -0.634025  0.993737   
8  1.133947 -1.066382 -1.486746  0.091896 -0.567492 -2.178004 -0.001447   

      acous      spch  
0 -0.683808 -0.402419  
1 -0.618050 -0.128577  
6 -0.223504 -0.265498  
7  0.499832  3.842137  
8  2.669839 -0.813183

### 2.4. Linear Regression Model with the Scaled Explanatory Variables

Next, fit a linear regression model with your `first genre dataset`, predicting `popularity` with each of your 9 *scaled song metric explanatory variables*.

In [15]:
import statsmodels.api as sm

X_dance_pop = genre1_scaled[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']]
y_dance_pop = genre1_data['popularity']

X_dance_pop_const = sm.add_constant(X_dance_pop)

model_dance_pop = sm.OLS(y_dance_pop, X_dance_pop_const).fit()

model_dance_pop.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.077
Model:                            OLS   Adj. R-squared:                  0.053
Method:                 Least Squares   F-statistic:                     3.231
Date:                Tue, 10 Oct 2023   Prob (F-statistic):           0.000878
Time:                        23:53:49   Log-Likelihood:                -1268.7
No. Observations:                 361   AIC:                             2557.
Df Residuals:                     351   BIC:                             2596.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         74.0332      0.434    170.612      0.000      73.180      74.887
bpm            0.3966      0.468      0.848      0.397      -0.523       1.316
nrgy          -1.9875      0.653     -3.046      0.002      -3.271      -0.704
dnce           0.5164      0.510      1.013      0.312      -0.486       1.519
dB             1.0528      0.594      1.773      0.077      -0.115       2.221
live          -0.7351      0.446     -1.647      0.100      -1.613       0.143
val            1.3090      0.523      2.503      0.013       0.281       2.338
dur            0.1151      0.453      0.254      0.800      -0.776       1.006
acous          0.4114      0.496      0.829      0.408      -0.565       1.388
spch          -0.7151      0.447     -1.600      0.110      -1.594       0.164
==============================================================================
Omnibus:                       16.317   Durbin-Watson:                   1.876
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               17.328
Skew:                          -0.529   Prob(JB):                     0.000173
Kurtosis:                       3.183   Cond. No.                         2.87
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 2.5. Top 3 Most Important Song Metrics

What are the three most important song metrics when it comes to predicting song popularity for the `genre 1` songs in this dataset with a linear regression model? Explain.

The importance of the song metrics in predicting song popularity in a linear regression model can be evaluated based on the absolute values of the coefficients and their statistical significance (p-values).

From the regression results:

1. **nrgy (Energy)**: Coefficient = -1.9875, p-value = 0.002
   - Energy is the most influential metric based on its coefficient's absolute value. The negative coefficient suggests that as the energy of a song increases, its popularity tends to decrease, at least for the "dance pop" genre in this dataset. The p-value indicates that this predictor is statistically significant.

2. **val (Valence)**: Coefficient = 1.3090, p-value = 0.013
   - Valence is the second most influential metric. A positive coefficient suggests that songs with a more positive mood are more popular. The statistical significance of valence is also confirmed by its p-value.

3. **dB (Decibel)**: Coefficient = 1.0528, p-value = 0.077
   - Decibel level comes third in terms of importance. The positive coefficient suggests that louder songs tend to be more popular. Although its p-value is slightly above the conventional 0.05 threshold for significance, it's close enough to be considered borderline significant.

### 2.6. Complicating Effect of Multicollinearity

Don't forget that when a linear regression model has issues with multicollinearity, our interpretations of the resulting slopes can be misleading and not reflect the relationships that exist between the explanatory variables and the response variable in our training dataset. This slope interpretation complication also holds for explanatory variables that have been scaled.

Does our linear regression model have an issue with multicollinearity? Explain.

To determine if our linear regression model has an issue with multicollinearity, we can look at a few indicators:

1. **Variance Inflation Factor (VIF)**: This is a common metric used to detect multicollinearity. If any variable has a VIF greater than 10 (some researchers use a threshold of 5), it's a sign that it's highly correlated with other predictors, indicating multicollinearity.
 
2. **Condition Number**: From the regression results, we see the "Cond. No." is 2.87. A condition number much greater than 30 is a sign of strong multicollinearity. In our case, the condition number is well below this threshold, suggesting that multicollinearity is not a significant issue.

3. **Coefficient Signs and Magnitudes**: If the signs or magnitudes of the coefficients don't align with what you'd expect based on domain knowledge or they change dramatically when a predictor is added/removed, it might suggest multicollinearity.

4. **Statistical Significance**: If several predictors are correlated, they might not be individually significant, but they could be significant when considered together.

In [16]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

X = genre1_scaled[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']]
X = sm.add_constant(X)

vif_data = pd.DataFrame()
vif_data["Variable"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

vif_data


Variable       VIF
0    const  1.000000
1      bpm  1.158123
2     nrgy  2.255317
3     dnce  1.376294
4       dB  1.867027
5     live  1.054645
6      val  1.448277
7      dur  1.086082
8    acous  1.304674
9     spch  1.057852

Given the VIF values presented:

- A VIF of 1 indicates that there's no correlation between a given predictor and the other predictors.
- A VIF between 1 and 5 suggests that there's a moderate correlation, but it's typically not severe enough to warrant corrective measures.
- A VIF above 5 can be problematic, as it indicates potential multicollinearity.

Looking at the VIF values for your predictors:

- All predictors have VIF values well below 5.
- The highest VIF is for `nrgy` at $2.255317$, which suggests a moderate correlation with other predictors but is not alarmingly high.

Given these results, we can conclude that multicollinearity is not an issue in this linear regression model for the genre 1 songs in your dataset. The predictors are not highly intercorrelated.

## 3. Second Genre Analysis

Next, we'd like to fit a *type of* linear regression model that predicts a songs `popularity` using the following explanatory variables for `just your genre 2 songs`:
* `bpm`	Beats Per Minute - The tempo of the song
* `nrgy`	Energy - How energetic the song is
* `dnce`	Danceability - How easy it is to dance to the song
* `dB`	Decibel - How loud the song is
* `live`	How likely the song is a live recording
* `val`	How positive the mood of the song is
* `dur`	Duration of the song
* `acous`	How acoustic the song is
* `spch`	The more the song is focused on spoken word

HOWEVER, we'd ALSO like to determine HOW IMPORTANT each of these explanatory variables are when it comes to predicting song `popularity`.


### 3.1 z-Score Scaling the Explanatory Variables

Next, scale each of your explanatory variables in your `second genre dataset` using the z-score scaling method that we discussed in class.

In [19]:
unique_genres = spotify_df['top genre'].unique()
print(unique_genres)

['dance pop' 'pop soul' 'atl hip hop' 'pop rap' 'big room'
 'canadian hip hop' 'disco house' 'romanian house' 'lilith'
 'detroit hip hop' 'asian american hip hop' 'east coast hip hop'
 'neo mellow' 'canadian pop' 'reggae fusion' 'idol' 'art pop' 'hip hop'
 'pop' 'talent show' 'modern alternative rock' 'indietronica' 'grime'
 'barbadian pop' 'acoustic pop' 'dutch house' 'belgian pop'
 'contemporary country' 'boy band' 'celtic rock' 'edm' 'indie rock'
 'australian dance' 'british soul' 'eau claire indie' 'dancefloor dnb'
 'permanent wave' 'hip pop' 'g funk' 'baroque pop' 'indie pop'
 'chicago rap' 'indie poptimism' 'french shoegaze' 'alternative metal'
 'indie folk' 'alternative rock' 'uk hip hop' 'electro house'
 'garage rock' 'israeli pop' 'alternative r&b' 'australian pop'
 'candy pop' 'modern rock' 'conscious hip hop' 'folk-pop'
 'alternative dance' 'k-pop' 'gangster rap' 'brostep' 'downtempo'
 'la indie' 'bass trap' 'metropopolis' 'electropop' 'electro'
 'destroy techno' 'emo' 'aust

In [21]:
genre2 = "pop"

genre2_songs = spotify_df[spotify_df['top genre'] == genre2]

std_dev_genre2 = genre2_songs[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']].std()

std_dev_genre2


bpm      30.869734
nrgy     13.895172
dnce     12.429601
dB        1.700472
live     11.753019
val      21.299746
dur      27.703417
acous    21.566575
spch      7.562439
dtype: float64

### 3.2 Linear Regression Model with the Scaled Explanatory Variables

Next, fit a linear regression model with your `second genre dataset`, predicting `popularity` with each of your 9 *scaled song metric explanatory variables*.

In [25]:
from scipy.stats import zscore
import statsmodels.formula.api as smf
# Filter out only the pop songs
genre2_songs = spotify_df[spotify_df['top genre'] == 'pop']

# Scaling the explanatory variables in genre2_songs dataset using z-score scaling
scaled_features_genre2 = genre2_songs[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']].apply(zscore)

# Adding the scaled features to the original genre2_songs DataFrame
scaled_genre2_songs = genre2_songs.copy()
scaled_genre2_songs[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']] = scaled_features_genre2

# Fitting a linear regression model to predict popularity using the scaled explanatory variables
model_genre2 = smf.ols("popularity ~ bpm + nrgy + dnce + dB + live + val + dur + acous + spch", data=scaled_genre2_songs).fit()
model_genre2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.234
Model:                            OLS   Adj. R-squared:                  0.088
Method:                 Least Squares   F-statistic:                     1.597
Date:                Wed, 11 Oct 2023   Prob (F-statistic):              0.144
Time:                        00:07:57   Log-Likelihood:                -188.81
No. Observations:                  57   AIC:                             397.6
Df Residuals:                      47   BIC:                             418.0
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     76.0000      0.969     78.443      0.000      74.051      77.949
bpm            0.5723      1.208      0.474      0.638      -1.858       3.003
nrgy          -3.1646      1.772     -1.786      0.081      -6.729       0.400
dnce          -0.4362      1.071     -0.407      0.685      -2.590       1.717
dB            -0.5591      1.585     -0.353      0.726      -3.747       2.629
live          -0.5893      1.051     -0.561      0.578      -2.704       1.525
val            2.6011      1.219      2.134      0.038       0.149       5.053
dur            0.9057      1.118      0.810      0.422      -1.343       3.155
acous          0.4364      1.277      0.342      0.734      -2.132       3.005
spch          -0.2739      1.180     -0.232      0.817      -2.648       2.101
==============================================================================
Omnibus:                        1.120   Durbin-Watson:                   1.495
Prob(Omnibus):                  0.571   Jarque-Bera (JB):                1.147
Skew:                          -0.308   Prob(JB):                        0.563
Kurtosis:                       2.679   Cond. No.                         3.83
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 3.3 Top 3 Most Important Song Metrics

What are the three most important song metrics when it comes to predicting song popularity for the `genre 2` songs in this dataset with a linear regression model? Explain.

To determine the importance of song metrics in the linear regression model, we often look at a combination of the magnitude of the coefficients (which represent the effect size) and their statistical significance.

From the regression results for the "pop" genre:

1. **`val` (Mood of the song)**: The coefficient for `val` is $2.6011$, and its p-value is 0.038, making it statistically significant at the $5\%$ level. This indicates that the mood of the song is the most important predictor among the given song metrics for predicting popularity in the "pop" genre. A one standard deviation increase in `val` is associated with an increase in the predicted popularity by approximately $2.6$ units, holding all other predictors constant.

2. **`nrgy` (Energy)**: The coefficient for `nrgy` is $-3.1646$, and its p-value is $0.081$. Although it's slightly above the common $0.05$ significance threshold, it's close enough to warrant consideration. The negative coefficient suggests that, on average, songs with higher energy might be slightly less popular in the "pop" genre, but this result isn't as strong as the one for `val`.

3. **`dB` (Decibel - Loudness of the song)**: The coefficient for `dB` is $-0.5591$, and its p-value is $0.726$. While the p-value suggests that the loudness is not a statistically significant predictor, the magnitude of the coefficient is relatively high compared to some other metrics. However, it's important to note that relying on the magnitude of coefficients without considering statistical significance can be misleading.

In summary, based on the given results, the most important song metric is `val` (Mood of the song). `nrgy` (Energy) might also play a role, but its effect is not as statistically robust. Other metrics like `dB` have coefficients with notable magnitudes, but they are not statistically significant in this model, so their importance is questionable.

### 3.4. Multicollinearity

Does this linear regression model have an issue with multicollinearity, thereby making us more skeptical of our slope interpretations? Explain.

To assess multicollinearity, we often rely on the Variance Inflation Factor (VIF). As a rule of thumb:
- A VIF of 1 indicates no multicollinearity.
- A VIF between 1 and 5 suggests moderate multicollinearity.
- A VIF greater than 5 indicates high multicollinearity.

From the VIF values earlier for the `pop` genre:


\begin{align*}
\text{bpm} & : 1.555332 \\
\text{nrgy} & : 3.345212 \\
\text{dnce} & : 1.220970 \\
\text{dB} & : 2.675681 \\
\text{live} & : 1.177005 \\
\text{val} & : 1.582483 \\
\text{dur} & : 1.331650 \\
\text{acous} & : 1.736861 \\
\text{spch} & : 1.484057 \\
\end{align*}


All of the VIF values are below 5, suggesting that there isn't a high degree of multicollinearity among the predictors in the model for the `pop` genre.

Thus, based on these VIF values, this linear regression model does not appear to have a significant issue with multicollinearity. Therefore, while we should always interpret regression results with caution, in this case, multicollinearity does not seem to be a major concern that would make us skeptical of our slope interpretations.

In [27]:
# Adjusting for the 'pop' genre
genre2_data = spotify_df[spotify_df['top genre'] == 'pop']

genre2_scaled = genre2_data.copy()
genre2_scaled[columns_to_scale] = (genre2_data[columns_to_scale] - genre2_data[columns_to_scale].mean()) / genre2_data[columns_to_scale].std()

# Extracting the scaled variables for the pop genre
X_genre2 = genre2_scaled[['bpm', 'nrgy', 'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch']]
# Adding a constant to the model (intercept)
X_genre2 = sm.add_constant(X_genre2)

# Calculating VIF for each variable
vif_genre2 = pd.DataFrame()
vif_genre2["Variable"] = X_genre2.columns
vif_genre2["VIF"] = [variance_inflation_factor(X_genre2.values, i) for i in range(X_genre2.shape[1])]

vif_genre2

Variable       VIF
0    const  1.000000
1      bpm  1.555332
2     nrgy  3.345212
3     dnce  1.220970
4       dB  2.675681
5     live  1.177005
6      val  1.582483
7      dur  1.331650
8    acous  1.736861
9     spch  1.484057

## 4. Fit Comparison

Which of these two genre linear regression models had a better fit? Explain.

Here's a summary of the R-squared and Adjusted R-squared values for both genres:

### Genre 1 (Dance Pop)
- R-squared: $0.077$
- Adjusted R-squared: $0.053$

### Genre 2 (Pop)
- R-squared: $0.234$
- Adjusted R-squared: $0.088$

### Analysis:
- **R-squared**: The model for the `pop` genre has a higher R-squared value (0.234) compared to the `dance pop` genre $(0.077)$. This means that the model for the `pop` genre explains more variance in the song popularity.
- **Adjusted R-squared**: The `pop` genre also leads with a higher Adjusted R-squared value (0.088) compared to the `dance pop` genre $(0.053)$. Adjusted R-squared is a more accurate measure as it takes the number of predictors into account, offering a more balanced comparison.

### Conclusion:
The linear regression model for the `pop` genre has a better fit as indicated by both higher R-squared and Adjusted R-squared values. This suggests that the model for the `pop` genre is more effective at explaining the variability in song popularity based on the included song metrics, at least within the scope of the training data provided.

Worked with Sophie Bustos and Ivan Mei on all parts; Please look at Sophie's version for the final.